network/netplan/netplan.go

```diff
@@ -76,41 +119,168 @@ type Netplan struct {
+	MinLinks           *int      `yaml:"min-links,omitempty"`
+	TransmitHashPolicy string    `yaml:"transmit-hash-policy,omitempty"`
+	ADSelect           IntString `yaml:"ad-select,omitempty"`
+	AllSlavesActive    *bool     `yaml:"all-slaves-active,omitempty"`
+	ARPInterval        *int      `yaml:"arp-interval,omitempty"`
+	ARPIPTargets       []string  `yaml:"arp-ip-targets,omitempty"`
+	ARPValidate        IntString `yaml:"arp-validate,omitempty"`
+	ARPAllTargets      IntString `yaml:"arp-all-targets,omitempty"`
+	UpDelay            *int      `yaml:"up-delay,omitempty"`
+	DownDelay          *int      `yaml:"down-delay,omitempty"`
+	FailOverMACPolicy  IntString `yaml:"fail-over-mac-policy,omitempty"`
+	// Netplan misspelled this as 'gratuitious-arp', not sure if it works with that name.
+	// We may need custom handling of both spellings.
+	GratuitousARP         *int      `yaml:"gratuitious-arp,omitempty"`
+	PacketsPerSlave       *int      `yaml:"packets-per-slave,omitempty"`
+	PrimaryReselectPolicy IntString `yaml:"primary-reselect-policy,omitempty"`
+	ResendIGMP            *int      `yaml:"resend-igmp,omitempty"`
+	// bonding.txt says that this can be a value from 1-0x7fffffff, should we be forcing it to be a hex value?
+	LearnPacketInterval *int   `yaml:"learn-packet-interval,omitempty"`
+	Primary             string `yaml:"primary,omitempty"`
+}
+
+// BridgeEthernetById takes a deviceId and creates a bridge with this device
 // using this devices config
 func (np *Netplan) BridgeEthernetById(deviceId string, bridgeName string) (err error) {
 	ethernet, ok := np.Network.Ethernets[deviceId]
 	if !ok {
-		return errors.NotFoundf("Device with id %q for bridge %q", deviceId, bridgeName)
+		return errors.NotFoundf("ethernet device with id %q for bridge %q", deviceId, bridgeName)
+	}
+	shouldCreate, err := np.shouldCreateBridge(deviceId, bridgeName)
+	if !shouldCreate {
+		// err may be nil, but we shouldn't continue creating
+		return errors.Trace(err)
+	}
+	np.createBridgeFromInterface(bridgeName, deviceId, &ethernet.Interface)
+	np.Network.Ethernets[deviceId] = ethernet
+	return nil
+}
+
+// BridgeVLANById takes a deviceId and creates a bridge with this device
+// using this devices config
+func (np *Netplan) BridgeVLANById(deviceId string, bridgeName string) (err error) {
+	vlan, ok := np.Network.VLANs[deviceId]
+	if !ok {
+		return errors.NotFoundf("VLAN device with id %q for bridge %q", deviceId, bridgeName)
+	}
+	shouldCreate, err := np.shouldCreateBridge(deviceId, bridgeName)
+	if !shouldCreate {
+		// err may be nil, but we shouldn't continue creating
+		return errors.Trace(err)
+	}
+	np.createBridgeFromInterface(bridgeName, deviceId, &vlan.Interface)
+	np.Network.VLANs[deviceId] = vlan
+	return nil
+}
+
+// BridgeBondById takes a deviceId and creates a bridge with this device
+// using this devices config
+func (np *Netplan) BridgeBondById(deviceId string, bridgeName string) (err error) {
```

Comment:
I might be tempted to create an abstraction for these 3, but perhaps passing in a device type string and a map pointer makes it less cogent...

Reply:
I pulled everything that wasn't type specific into helper functions. I couldn't think of a better way to simplify them, though I would consider switching away from individual BridgeFoo and just have a BridgeDevice that knows about all three types.
Thoughts?